README.md
README for GNU indent 2.2.12
============================

  GNU indent changes the appearance of a C program by inserting or
  deleting whitespace according to a plethora of options.  Some
  canned styles of formatting are supported as well.  GNU indent is
  a descendant of BSD indent.  This file is part of the GNU indent
  distribution.

  GNU indent does NOT work for C++, if it does for you then you're
  just lucky.  Don't expect support for C++.

  Read the file NEWS for more information, especially to see what's
  different from the last version, and what future versions may do.

  To install indent, please read the file INSTALL included with the
  distribution. If you are installing indent on VMS, see the file
  miscel/README.VMS. For VMS and other platforms there is a variety 
  of files in the directory miscel that may be of use. Move the
  files that may be of use to the src directory before use.

  For license and copying information, see the file COPYING.

Documentation
-------------

  For complete documentation on the Web, please visit:

     http://www.gnu.org/software/indent/

  GNU uses a language called "texinfo" for its documentation source
  which produces both an on-line, interactive version and a hardcopy
  version for printing.  An HTML version is now produced as well.

  The on-line versions, `indent.info` and `indent.1`, are installed
  along with the executable by typing "make install".
  Refer to the Texinfo manual for more information on the `info` system.

  The hardcopy version is produced from the source file `indent.texinfo`
  in a pdf or Postscript format file.  To produce the pdf
  version type:

    make pdf

  To produce the Postscript version, type:

    make ps

  Note that you must have the TeX typographical system installed for
  this to work.

  The HTML version is included along with the distribution, and
  consists of the file:

     doc/indent.html

  This file can also be accessed at the URL 
        http://www.gnu.org/software/indent/manual/

  To create the HTML version from the texinfo version, type:

     make html

  Obviously you will need the texi2html tool.

Bug reports
-----------

  Please address bug-reports and suggestions or comments to:

    bug-indent@gnu.org

  How to report bugs:

  Please begin your bug report with a concise and factual description
  of the behaviour you consider to be abberant.  If you have narrowed
  the problem down to the use of a particular option, please mention
  that.  In general, also be sure to mention the command line options
  used to invoke indent, as well as the contents of your .indent.pro
  if you are using one.

  Then include the *smallest possible* piece of code which generates
  the described symptom.  It is very helpful if you can manage to
  reduce the size of the sample code.

  Finally, if you have investigated the problem further, include
  debugger stack traces, modifications, or speculation you have.

Sending code
------------

  If you send a patch as part of your bug fix, or send me code for new
  features, please send them as diffs.  In particular, include the
  result of "diff -c <old_file> <new_file>".

  Because of the regression testing the project requires of indent, 
  it is extremely unlikely that your code will be installed into indent
  "as-is";  however it may still be helpful, and will likely have
  solved your particular problem.


ChangeLog
2018-09-02 Andrej Shadura <andrew@shadura.me>
	* Add C99 and C11 keywords; add typeof.
	* Don’t insert a space between ! in !!x
	* Don’t insert a space after __attribute__ when -bs is disabled
	* Include -nbs in -linux

2018-09-02 Helmut Grohne <helmut@subdivi.de>
	* Use the build architecture compiler for build tools

2018-09-02 Hanno Böck <hanno@hboeck.de>
	* Fix 46626: Invalid memory reads in parse()
	* Fix 46627: Out of bounds read when using indent on a zero-byte file

2018-09-01 Andrej Shadura <andrew@shadura.me>
	* Update translations.
	* Update headers and autogenerated code.
	* Stop using texi2{pdf,html}, use makeinfo instead.
	* Transcode most of translations to UTF-8.
	* Mark legitimate switch fall though cases.

2015-10-11 Petr Písař <ppisar@redhat.com>
	* Fix -nbdfa and -nbdfe typo.
	* Return non-zero exit code on tests failure.
	* Support hexadecimal floats.

2015-08-27 Andrej Shadura <andrew@shadura.me>
	* Fix format strings in translations

2015-08-25 Andrej Shadura <andrew@shadura.me>
	* Stop translating debug strings.
	* Update the translations.
	* Drop support for VMS and Watcom.

2015-08-24 Andrej Shadura <andrew@shadura.me>
	* Update the translations.
	* Update autoconf files.

2015-08-21 Andrej Shadura <andrew@shadura.me>
	* Reformat the README file using basic Markdown.
	* Regenerate headers.
	* Regenerate documentation.
	* Regenerate translations.

2015-07-04 Tim Hentenaar <tim@hentenaar.com>
	* handletoken.c: Detect C99 compound literals.
	* Makefile.am: Add -Wextra to AM_CFLAGS.
	* output.c: For -as, if we're in an if/for/while statement, and
	  we're breaking the line somewhere within the parenthesized
	  expression, use the same indent level as the parent statement.

2015-06-30 Tim Hentenaar <tim@hentenaar.com>
	* Tighten-up the CFLAGS with additional warning flags.
	* Ensure indent builds with clang.
	* Fix various compiler warnings.

2015-06-17 Tim Hentenaar <tim@hentenaar.com>
	* comments.c: CHECK_COM_SIZE: Use unsigned type for size
	* comments.c: CHECK_COM_SIZE: Add a parameter for the expected size
	  we're about to write.
	* comments.c: Prevent heap corruption via writing past the end of
	  combuf by adding a size check in places where we're writing to
	  combuf by looping and incrememnting e_com.
	* indent.c: Don't force a newline after an rbrace and before a
	  comma in a block initializer.
	* Added an option (-ntac/--dont-tab-align-comments) which if specified
	  prevents indent from automagically pushing comments out to the
	  nearest tabstoo.
	* Fixed recognition of function pointers with multiple levels of
	  indirection by the -npcs stuff.
	* Add an 'install' target to the regression Makefile so that
	  'make install' doesn't fail.

2015-06-16 Tim Hentenaar <tim@hentenaar.com>
	* Use want_blank instead of copying a space to e_code to avoid
	  double spaces.
	* Fix print_comment reading past the end of the buffer when reading
	  less than a full comment from stdin.
	* Ensure that there's only one point where exit(3) is used.
	* Add xfree() to globs.
	* xfree() common heap-allocated storage.
	* Fixed some memory leaks in the profile stuff.
	* Be sure to initialize the two automatic storage buffers used in
	  scan_profile(). This fixes a warnings from valgrind.

2015-06-15 Tim Hentenaar <tim@hentenaar.com>
	* Added -par to -kr.
	* Fixed detection of struct initializer lists as block_init.
	* Added an option (-sar/--spaces-around-initializers) to add a space
	  after the lbrace and before the rbrace of struct initializers.

2015-06-14 Tim Hentenaar <tim@hentenaar.com>
	* Added an option (-as/--align-with-spaces) which if indenting with
	  tabs will ident to the current indentation level, and then pad
	  with spces up till the target_column.
	* If -npcs is set, don't add a space between a function pointer's
	  name and its arglist.
	* Fix indentation of block comments following braces.

2015-06-13 Tim Hentenaar <tim@hentenaar.com>
	* Fix regression introduced in hg revision 25c27d429590
	* Add an option (-slc/--single-line-conditionals) to allow
	  conditionals and their inner statement to be on the same line if
	  they're not braced. This should fix GNU bug #42357.
	* Added a small patch to add size_t, wchar_t and ptrdiff_t to
	  user_specials automatically.

2014-04-07  Andrew Shadura  <andrew@shadura.me>
	* Don't add extra 'const' keyword to C++ programs (closes Debian bugs
	  #507173 and #349065).

2014-03-29 Daniel P. Valentine <dpv@world.std.com>
	* Added option -gts to treat _("...") and N_("...") as strings rather than
	  functions with string arguments. (args.c, indent.h, lexi.c, indent.texi,
	  regression/input/gettext-strings.c)

2014-03-29  Andrew Shadura  <andrew@shadura.me>
	* Add an option to fix nested comments (-fnc/--fix-nested-comment).
	  This option which tries to fix nested comments by replacing '/*'
	  inside comments by a space followed by '*'. By default, this option
	  is turned off.
	* Pass -npro to indent when running tests.

2014-03-28  gettextize  <bug-gnu-gettext@gnu.org>

	* m4/gettext.m4: Upgrade to gettext-0.18.3.
	* m4/iconv.m4: Upgrade to gettext-0.18.3.
	* m4/lib-ld.m4: Upgrade to gettext-0.18.3.
	* m4/lib-link.m4: Upgrade to gettext-0.18.3.
	* m4/lib-prefix.m4: Upgrade to gettext-0.18.3.
	* m4/nls.m4: New file, from gettext-0.18.3.
	* m4/po.m4: New file, from gettext-0.18.3.
	* m4/progtest.m4: Upgrade to gettext-0.18.3.
	* m4/codeset.m4: Upgrade to gettext-0.18.3.
	* m4/fcntl-o.m4: New file, from gettext-0.18.3.
	* m4/glibc2.m4: New file, from gettext-0.18.3.
	* m4/glibc21.m4: Upgrade to gettext-0.18.3.
	* m4/intdiv0.m4: Upgrade to gettext-0.18.3.
	* m4/intl.m4: New file, from gettext-0.18.3.
	* m4/intldir.m4: New file, from gettext-0.18.3.
	* m4/intlmacosx.m4: New file, from gettext-0.18.3.
	* m4/intmax.m4: New file, from gettext-0.18.3.
	* m4/inttypes_h.m4: Upgrade to gettext-0.18.3.
	* m4/inttypes-pri.m4: Upgrade to gettext-0.18.3.
	* m4/lcmessage.m4: Upgrade to gettext-0.18.3.

src/gperf.c
/* ANSI-C code produced by gperf version 3.1 */
/* Command-line: gperf -D -c -l -p -t -T -g -j1 -o -K rwd -L ANSI-C -N is_reserved indent.gperf  */
/* Computed positions: -k'2-3' */

#if !((' ' == 32) && ('!' == 33) && ('"' == 34) && ('#' == 35) \
      && ('%' == 37) && ('&' == 38) && ('\'' == 39) && ('(' == 40) \
      && (')' == 41) && ('*' == 42) && ('+' == 43) && (',' == 44) \
      && ('-' == 45) && ('.' == 46) && ('/' == 47) && ('0' == 48) \
      && ('1' == 49) && ('2' == 50) && ('3' == 51) && ('4' == 52) \
      && ('5' == 53) && ('6' == 54) && ('7' == 55) && ('8' == 56) \
      && ('9' == 57) && (':' == 58) && (';' == 59) && ('<' == 60) \
      && ('=' == 61) && ('>' == 62) && ('?' == 63) && ('A' == 65) \
      && ('B' == 66) && ('C' == 67) && ('D' == 68) && ('E' == 69) \
      && ('F' == 70) && ('G' == 71) && ('H' == 72) && ('I' == 73) \
      && ('J' == 74) && ('K' == 75) && ('L' == 76) && ('M' == 77) \
      && ('N' == 78) && ('O' == 79) && ('P' == 80) && ('Q' == 81) \
      && ('R' == 82) && ('S' == 83) && ('T' == 84) && ('U' == 85) \
      && ('V' == 86) && ('W' == 87) && ('X' == 88) && ('Y' == 89) \
      && ('Z' == 90) && ('[' == 91) && ('\\' == 92) && (']' == 93) \
      && ('^' == 94) && ('_' == 95) && ('a' == 97) && ('b' == 98) \
      && ('c' == 99) && ('d' == 100) && ('e' == 101) && ('f' == 102) \
      && ('g' == 103) && ('h' == 104) && ('i' == 105) && ('j' == 106) \
      && ('k' == 107) && ('l' == 108) && ('m' == 109) && ('n' == 110) \
      && ('o' == 111) && ('p' == 112) && ('q' == 113) && ('r' == 114) \
      && ('s' == 115) && ('t' == 116) && ('u' == 117) && ('v' == 118) \
      && ('w' == 119) && ('x' == 120) && ('y' == 121) && ('z' == 122) \
      && ('{' == 123) && ('|' == 124) && ('}' == 125) && ('~' == 126))
/* The character set is not based on ISO-646.  */
#error "gperf generated tables don't work with this execution character set. Please report a bug to <bug-gperf@gnu.org>."
#endif


#define TOTAL_KEYWORDS 32
#define MIN_WORD_LENGTH 2
#define MAX_WORD_LENGTH 8
#define MIN_HASH_VALUE 3
#define MAX_HASH_VALUE 41
/* maximum key range = 39, duplicates = 0 */

#if defined(__GNUC__) && !defined(__clang__)
__inline
#else
#ifdef __cplusplus
inline
#endif
#endif
static unsigned int
hash (register const char *str, register size_t len)
{
  static unsigned char asso_values[] =
    {
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 15, 42, 12, 42, 42,
      42,  9, 25,  9,  8,  7, 42, 42,  0, 42,
       5,  1, 13, 42, 15,  0,  0, 16, 42, 18,
      24, 12, 15, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42, 42, 42, 42, 42,
      42, 42, 42, 42, 42, 42
    };
  register unsigned int hval = len;

  switch (hval)
    {
      default:
        hval += asso_values[(unsigned char)str[2]];
      /*FALLTHROUGH*/
      case 2:
        hval += asso_values[(unsigned char)str[1]];
        break;
    }
  return hval;
}

templ_ty *
is_reserved (register const char *str, register size_t len)
{
  static unsigned char lengthtable[] =
    {
       2,  4,  4,  5,  6,  3,  8,  4,  5,  4,  8,  5,  6,  4,
       5,  6,  3,  5,  6,  6,  6,  4,  4,  8,  2,  6,  5,  6,
       6,  7,  6,  7
    };
  static templ_ty wordlist[] =
    {
#line 33 "indent.gperf"
      {"do", rw_sp_nparen,},
#line 32 "indent.gperf"
      {"else", rw_sp_else,},
#line 27 "indent.gperf"
      {"goto", rw_break,},
#line 13 "indent.gperf"
      {"float", rw_decl,},
#line 22 "indent.gperf"
      {"global", rw_decl,},
#line 10 "indent.gperf"
      {"int", rw_decl,},
#line 25 "indent.gperf"
      {"volatile", rw_decl,},
#line 15 "indent.gperf"
      {"long", rw_decl,},
#line 12 "indent.gperf"
      {"const", rw_decl,},
#line 24 "indent.gperf"
      {"void", rw_decl,},
#line 18 "indent.gperf"
      {"unsigned", rw_decl,},
#line 16 "indent.gperf"
      {"short", rw_decl,},
#line 28 "indent.gperf"
      {"return", rw_return,},
#line 4 "indent.gperf"
      {"case", rw_case,},
#line 7 "indent.gperf"
      {"union", rw_struct_like,},
#line 21 "indent.gperf"
      {"static", rw_decl,},
#line 31 "indent.gperf"
      {"for", rw_sp_paren,},
#line 30 "indent.gperf"
      {"while", rw_sp_paren,},
#line 6 "indent.gperf"
      {"struct", rw_struct_like,},
#line 19 "indent.gperf"
      {"signed", rw_decl,},
#line 14 "indent.gperf"
      {"double", rw_decl,},
#line 11 "indent.gperf"
      {"char", rw_decl,},
#line 8 "indent.gperf"
      {"enum", rw_enum,},
#line 20 "indent.gperf"
      {"register", rw_decl,},
#line 29 "indent.gperf"
      {"if", rw_sp_paren,},
#line 34 "indent.gperf"
      {"sizeof", rw_sizeof,},
#line 5 "indent.gperf"
      {"break", rw_break,},
#line 23 "indent.gperf"
      {"extern", rw_decl,},
#line 3 "indent.gperf"
      {"switch", rw_switch,},
#line 17 "indent.gperf"
      {"typedef", rw_decl,},
#line 26 "indent.gperf"
      {"va_dcl", rw_decl,},
#line 9 "indent.gperf"
      {"default", rw_case,}
    };

  static signed char lookup[] =
    {
      -1, -1, -1,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10,
      11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
      25, 26, 27, 28, 29, 30, -1, -1, -1, -1, -1, -1, -1, 31
    };

  if (len <= MAX_WORD_LENGTH && len >= MIN_WORD_LENGTH)
    {
      register unsigned int key = hash (str, len);

      if (key <= MAX_HASH_VALUE)
        {
          register int idx = lookup[key];

          if (idx >= 0)
            {
              if (len == lengthtable[idx])
                {
                  register const char *s = wordlist[idx].rwd;

                  if (*str == *s && !memcmp (str + 1, s + 1, len - 1))
                    return &wordlist[idx];
                }
            }
        }
    }
  return 0;
}
